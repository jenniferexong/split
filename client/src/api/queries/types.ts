import { ApiProduct, ApiReceipt, ApiPerson, ApiStore } from 'api';

export type QueryResult<T> = () => Promise<T>;

export interface ProductResult {
  product: ApiProduct;
}

export interface ProductsResult {
  products: ApiProduct[];
}

export interface StoreResult {
  store: ApiStore;
}

export interface StoresResult {
  stores: ApiStore[];
}

export interface PeopleResult {
  people: ApiPerson[];
}

export interface PersonResult {
  person: ApiPerson;
}

export interface ReceiptResult {
  receipt: ApiReceipt;
}

export interface DeleteReceiptResult {
  deleteReceipt: ApiReceipt;
}

export interface ReceiptsResult {
  receipts: ApiReceipt[];
}
