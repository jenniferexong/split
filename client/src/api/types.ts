export interface ApiProduct {
  id: number;
  name: string;
}

export interface ApiStore {
  id: number;
  name: string;
}

export interface ApiPerson {
  id: number;
  firstName: string;
  lastName: string;
  email: string;
}

// TODO
export interface ApiReceipt {
  id: number;
}
