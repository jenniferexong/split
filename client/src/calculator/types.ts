import { ApiPerson, ApiProduct, ApiStore } from 'api';

export type Whose = 'mine' | 'split' | 'theirs';

export interface SplitType {
  person: ApiPerson;
  antecedent: number;
}

export interface ItemType {
  product: ApiProduct | undefined;
  splits: SplitType[];
  price: number;
}

export interface ReceiptType {
  store: ApiStore | undefined;
  date: Date | undefined;
  items: ItemType[];
  subtotal: number;
}

export interface PersonType {
  person?: ApiPerson;
  receipts: ReceiptType[];
  image?: string;
}

export interface AppType {
  people: PersonType[];
}

export interface InvoiceData {
  person: ApiPerson | undefined;
  totalSpendings: number;
  actualSpendings: number;
  oweings: number;
}
