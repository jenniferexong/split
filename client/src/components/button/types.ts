import { ReactNode } from 'react';

export interface BaseButtonProps {
  /** Button text */
  children: ReactNode;
  onClick?: () => void;
}
