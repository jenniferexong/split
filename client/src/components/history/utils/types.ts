export interface PersonTotal {
  name: string;
  total: number;
}

// Indexed by person id
export type PersonTotals = Record<number, PersonTotal>;
