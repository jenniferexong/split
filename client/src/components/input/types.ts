import { ApiPerson, ApiProduct, ApiStore } from '../../api';

export interface Option<T> {
  label: string;
  data: T;
}

export type StoreOption = Option<ApiStore>;

export type ProductOption = Option<ApiProduct>;

export type PersonOption = Option<ApiPerson>;
