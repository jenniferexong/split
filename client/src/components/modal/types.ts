import { ReactElement, ReactNode } from 'react';

export interface BaseModalProps {
  show: boolean;
  handleClose: () => void;
  footerContent?: ReactElement;
  children: ReactNode | ReactNode[];
}

export type ModalProps = Pick<BaseModalProps, 'show' | 'handleClose'>;
