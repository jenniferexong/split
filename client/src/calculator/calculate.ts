import { unreachable } from '../utils/unreachable';
import { AppType, InvoiceData } from './types';

interface CalculateResult {
  globalTotal: number;
  /** In the same order of people in the app */
  invoices: InvoiceData[];
}

export const calculate = (data: AppType): CalculateResult => {
  let globalTotal = 0;

  const invoices: InvoiceData[] = [
    {
      totalSpendings: 0,
      actualSpendings: 0,
      oweings: 0,
    },
    {
      totalSpendings: 0,
      actualSpendings: 0,
      oweings: 0,
    },
  ];

  data.people.forEach((person, personIndex) => {
    person.receipts.forEach(receipt => {
      receipt.items.forEach(item => {
        const { price, whose } = item;

        globalTotal += price;
        invoices[personIndex].actualSpendings += price;

        switch (whose) {
          case 'mine': {
            invoices[personIndex].totalSpendings += price;
            break;
          }
          case 'theirs': {
            invoices[personIndex ^ 1].totalSpendings += price;
            invoices[personIndex ^ 1].oweings += price;
            invoices[personIndex].oweings -= price;
            break;
          }
          case 'split': {
            const splitCost = price / 2;
            invoices[personIndex].totalSpendings += splitCost;
            invoices[personIndex ^ 1].totalSpendings += splitCost;
            invoices[personIndex ^ 1].oweings += splitCost;
            invoices[personIndex].oweings -= splitCost;
            break;
          }
          default: {
            unreachable(whose);
          }
        }
      });
    });
  });

  invoices.forEach(invoice => {
    invoice.oweings = invoice.oweings < 0 ? 0 : invoice.oweings;
  });

  return { globalTotal, invoices };
};
