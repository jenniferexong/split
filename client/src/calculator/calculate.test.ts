import { calculate } from './calculate';
import {
  adam,
  apples,
  bob,
  carrots,
  countdown,
  exampleReceipt,
  steak,
} from './exampleData';
import { EntryData } from './types';

describe('calculate', () => {
  const data: EntryData = {
    people: [
      {
        person: adam,
        receipts: [
          {
            date: new Date('01-01-22'),
            store: countdown,
            items: [
              {
                product: apples,
                splits: [
                  { person: adam, antecedent: 1 },
                  { person: bob, antecedent: 1 },
                ],
                price: 550,
              },
              {
                product: apples,
                splits: [
                  { person: adam, antecedent: 1 },
                  { person: bob, antecedent: 1 },
                ],
                price: 440,
              },
              {
                product: apples,
                splits: [{ person: adam, antecedent: 1 }],
                price: 20.5,
              },
              {
                product: carrots,
                splits: [{ person: bob, antecedent: 1 }],
                price: 21.8,
              },
            ],
            sequence: 0,
          },
        ],
      },
      {
        person: bob,
        receipts: [
          {
            date: new Date('01-01-22'),
            store: countdown,
            items: [
              {
                product: steak,
                splits: [{ person: bob, antecedent: 1 }],
                price: 4.56,
              },
              {
                product: apples,
                splits: [{ person: adam, antecedent: 1 }],
                price: 2.69,
              },
              {
                product: carrots,
                splits: [{ person: bob, antecedent: 1 }],
                price: 1.5,
              },
              {
                product: apples,
                splits: [{ person: bob, antecedent: 1 }],
                price: 10.4,
              },
              {
                product: carrots,
                splits: [{ person: adam, antecedent: 1 }],
                price: 1.4,
              },
              {
                product: apples,
                splits: [
                  { person: adam, antecedent: 1 },
                  { person: bob, antecedent: 1 },
                ],
                price: 30.5,
              },
            ],
            sequence: 1,
          },
        ],
      },
    ],
  };

  describe('calculates the correct information', () => {
    test('case 1', () => {
      const { globalTotal, invoices } = calculate(data);

      expect(globalTotal).toBeCloseTo(1083.35);

      expect(invoices[0].totalSpendings).toBeCloseTo(534.84);
      expect(invoices[0].actualSpendings).toBeCloseTo(1032.3);
      expect(invoices[0].owings).toBe(0);

      expect(invoices[1].totalSpendings).toBeCloseTo(548.51);
      expect(invoices[1].actualSpendings).toBeCloseTo(51.05);
      expect(invoices[1].owings).toBeCloseTo(497.46);
      expect({ globalTotal, invoices }).toMatchSnapshot();
    });

    test('case 2', () => {
      const exampleData: EntryData = {
        people: [
          {
            person: adam,
            receipts: [exampleReceipt(adam, bob)],
          },
          {
            person: bob,
            receipts: [exampleReceipt(bob, adam)],
          },
        ],
      };

      const { globalTotal, invoices } = calculate(exampleData);

      expect(globalTotal).toBeCloseTo(38.3);
      expect(invoices[0].totalSpendings).toBeCloseTo(19.15);
      expect(invoices[0].actualSpendings).toBeCloseTo(19.15);
      expect(invoices[0].owings).toBe(0);

      expect(invoices[1].totalSpendings).toBeCloseTo(19.15);
      expect(invoices[1].actualSpendings).toBeCloseTo(19.15);
      expect(invoices[1].owings).toBe(0);

      expect({ globalTotal, invoices }).toMatchSnapshot();
    });
  });
});
