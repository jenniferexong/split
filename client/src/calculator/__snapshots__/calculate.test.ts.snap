// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculate calculates the correct information case 1 1`] = `
Object {
  "globalTotal": 1083.3500000000001,
  "invoices": Array [
    Object {
      "actualSpendings": 1032.3,
      "oweings": 0,
      "person": Object {
        "email": "adam.agar@email.com",
        "firstName": "Adam",
        "id": 1,
        "lastName": "Agar",
      },
      "totalSpendings": 534.84,
    },
    Object {
      "actualSpendings": 51.05,
      "oweings": 497.4599999999999,
      "person": Object {
        "email": "bob.luza@email.com",
        "firstName": "Bob",
        "id": 2,
        "lastName": "luza",
      },
      "totalSpendings": 548.5099999999999,
    },
  ],
}
`;

exports[`calculate calculates the correct information case 2 1`] = `
Object {
  "globalTotal": 38.3,
  "invoices": Array [
    Object {
      "actualSpendings": 19.15,
      "oweings": 0,
      "person": Object {
        "email": "adam.agar@email.com",
        "firstName": "Adam",
        "id": 1,
        "lastName": "Agar",
      },
      "totalSpendings": 19.15,
    },
    Object {
      "actualSpendings": 19.15,
      "oweings": 0,
      "person": Object {
        "email": "bob.luza@email.com",
        "firstName": "Bob",
        "id": 2,
        "lastName": "luza",
      },
      "totalSpendings": 19.15,
    },
  ],
}
`;
